source/How-To-Guides/Launch-file-different-formats.rst

```diff
@@ -10,12 +10,321 @@ Using Python, XML, and YAML for ROS 2 Launch Files
    :local:
 
 ROS 2 launch files can be written in Python, XML, and YAML.
-This guide shows how to use these different formats to accomplish the same task, as well as has some discussion on when to use each format.
+This guide shows how to use these different formats to accomplish the same task. It also discusses when to use each format.
 
-Launch file examples
---------------------
+Examples
+--------
+
+Taking arguments
+^^^^^^^^^^^^^^^^
+
+Each launch file performs the following actions:
+
+* Declares arguments with and without defaults.
+* Logs a message that interpolates those arguments.
+
+.. tabs::
+
+   .. group-tab:: Python
+
+      .. code-block:: python
+
+        # example_launch.py
+
+        from launch import LaunchDescription
+        from launch.actions import DeclareLaunchArgument
+        from launch.actions import LogInfo
+        from launch.substitutions import LaunchConfiguration
```

Comment:
Maybe not the right place for this but I don't really understand why `LaunchConfiguration` is part of substitutions or what substitutions are anyway.  The [original roslaunch documentation](https://wiki.ros.org/roslaunch/XML#substitution_args) is pretty clear about this - I haven't found anything nearly so clear for ROS2 (and note that that part of the design doc is a TODO!) and it seems like ROS2 extends the definition.  [This tutorial](https://docs.ros.org/en/foxy/Tutorials/Intermediate/Launch/Using-Substitutions.html#:~:text=Substitutions%20are%20variables%20that%20are,in%20ROS%202%20launch%20files.) starts off by defining substitutions as a tool for re-using launch files but I can't figure out if they're variables or functions or some combination of the two?  The tutorial gives a bunch of examples that are text manipulation, but it doesn't make it clear, however, why substitutions are the right choice for this - why can't I just do normal Python string manipulation?  Should all text manipulation go through a substitution?  If so, why was it designed this way?

A description of all the available substitutions and how to use them would be very helpful.  I was unable to find one other than by reading the source code.

Reply:
>  can't figure out if they're variables or functions or some combination of the two?

It's easier to think about them as expressions evaluated in `launch` context.

> why substitutions are the right choice for this - why can't I just do normal Python string manipulation?

Because only substitutions can "easily" access `launch` context, where runtime state lives.